src/main/java/com/openlattice/authorization/EdmAuthorizationHelper.java

```diff
@@ -112,15 +112,99 @@ public EdmAuthorizationHelper( EdmManager edm, AuthorizationManager authz ) {
+                    .values().stream()
+                    .reduce( ( properties, nextProperties ) ->
+                            Maps.difference( properties, nextProperties ).entriesInCommon() )
+                    .orElse( Maps.newHashMap() );
+
+            return linkingEntitySet.getLinkedEntitySets().stream().collect( Collectors.toMap(
+                    Function.identity(),
+                    esId -> authorizedProperties
+            ) );
+        }
+    }
+
+    /**
+     * Note: entitysets are assumed to have same entity type
+     */
+    public Map<UUID, Map<UUID, PropertyType>> getAuthorizedPropertyTypesByNormalEntitySet(
+            EntitySet linkingEntitySet,
+            Set<UUID> selectedProperties,
+            EnumSet<Permission> requiredPermissions ) {
+        if ( linkingEntitySet.getLinkedEntitySets().isEmpty() ) {
+            return Maps.newHashMap();
+        } else {
+            final var authorizedPropertyTypesOfNormalEntitySets = getAuthorizedPropertyTypes(
+                    linkingEntitySet.getLinkedEntitySets(),
+                    selectedProperties,
+                    requiredPermissions,
+                    Principals.getCurrentPrincipals() );
+
+            final var authorizedProperties = authorizedPropertyTypesOfNormalEntitySets
+                    .values().stream()
+                    .reduce( ( properties, nextProperties ) ->
+                            Maps.difference( properties, nextProperties ).entriesInCommon() )
+                    .orElse( Maps.newHashMap() );
+
+            return linkingEntitySet.getLinkedEntitySets().stream().collect( Collectors.toMap(
+                    Function.identity(),
+                    esId -> authorizedProperties
+            ) );
+        }
+    }
+
+    public Map<UUID, PropertyType> getAuthorizedPropertyTypesOfLinkingEntitySet(
+            EntitySet linkingEntitySet,
+            EnumSet<Permission> requiredPermissions ) {
+        return getAuthorizedPropertyTypesOfLinkingEntitySet(
+                linkingEntitySet, requiredPermissions, Principals.getCurrentPrincipals() );
+    }
+
+    public Map<UUID, PropertyType> getAuthorizedPropertyTypesOfLinkingEntitySet(
+            EntitySet linkingEntitySet,
+            EnumSet<Permission> requiredPermissions,
+            Set<Principal> principals ) {
+        if ( linkingEntitySet.getLinkedEntitySets().isEmpty() ) {
+            return Maps.newHashMap();
+        } else {
+            final var authorizedPropertyTypesOfNormalEntitySets = getAuthorizedPropertiesOnEntitySets(
+                    linkingEntitySet.getLinkedEntitySets(), requiredPermissions, principals );
+
+            return authorizedPropertyTypesOfNormalEntitySets
+                    .values().stream()
+                    .reduce( ( properties, nextProperties ) ->
+                            Maps.difference( properties, nextProperties ).entriesInCommon() )
```

Comment:
Does this account for partial permission matches? i.e. if a linking entity set has 2 normal entity sets (es1, es2) and you have READ and WRITE on [es1, pt1], and only READ on [es2, pt1], will this set difference match and filter permissions to READ only?

Reply:
Good catch!

Reply:
Nope, it just returns property types for which the user has the required permissions on all normal entity sets. The example what you described can be achieved with 2 separate calls, 1 for checking WRITE and 1 for READ.

Reply:
ah true -- I was thinking of the get permissions calls in HazelcastAuthorizationService (btw is that set up for the new linked entity set permissions model?)